### ISSUE TITLE
Please give me a way to skip tools download

### ISSUE THREAD

**Issue by cweagans** [docksal/docksal#1258]
I don't want a random shell script downloading executables to my machine. Is there some way to not download docker, docker-compose, and docker-machine when I run `fin update`? Docker for Mac already provides these tools.

**Comment by achekulaev** [comment]
1. Yes there is a way. If you look at the `install_tools_mac` function in the `fin` script, you will notice that it only installs binary tools if currently installed tools' versions (`docker`, `docker-compose`, `docker-machine`) do not meet the requirements. So if your tools are of the same or higher than the required version, then nothing will get downloaded to your workstation.

2. Even when binaries are downloaded, they are not installed by PATH. They are installed into `~/.docksal/bin` and are only used by `fin`. Your original tools are not touched and running `docker-compose` from the terminal will still launch your original version, while running `fin docker-compose ...` will run the version fin has installed for itself into the aforementioned folder.

**Why we download tools to begin with**

Initially we used not to download them, but then users were coming back with weird Docker, Docker Compose or Docker machine errors. Those errors spawned from weird combinations of modern and outdated tools compatibility and also from compatibility of `fin` and those tools.

**Comment by cweagans** [comment]
I'm currently running Docker Desktop 2.1.0.5, which comes with Docker 19.03.5. `fin` installed 19.03.4. Regardless of where these tools are installed, I'd like a way to completely avoid having random executables downloaded to my machine.

**Comment by achekulaev** [comment]
Could you explain the principal difference of Docksal installer downloading random executables and say `brew` package installator downloading random executables? 

Imagine I finally found time and made a custom brew package for Docksal, and imagine you could now `brew cask install docksal`. 

Do you really go and check which "random" executables every brew package included?

Is there a way to completely avoid brew installing random executables for a package if you already have ones?

**Comment by cweagans** [comment]
I expect brew to download executables, and when it does, I've explicitly asked it to do so and it confirms if it needs to download dependencies and such. It also verifies that what it downloaded is what it expected to download using sha256 hashes in the formula.

I do not expect `fin` to download executables whenever it feels like it, and given that a) I did not ask it to download executables and b) I have no way of knowing specifically what was downloaded, from where, etc and c) I do not want to use those executables, it stands to reason that there should be a way to opt-out of that process.

**Comment by achekulaev** [comment]
I will definitely add checking for sha256 hashes, this is a very solid remark. I can also make `fin` issue a warning that it is going to download those binaries. All this is doable and if you would like to make this the point of this ticket.

`fin` downloads executables during Docksal installation and `fin update` commands, not when it  "feels like it". User has to issue that command first, or agree to the update when `fin` offers it.

However not downloading them at all might be problematic to implement, because it is a very narrow case: when your OS is macOS, when your Docker is Docker Desktop, when all your software versions are correct, then there is no need to install tools. In all other cases you do need those tools for `fin` to actually work.

**Comment by cweagans** [comment]
I would argue that DfM + macOS is not the only case where this is important. For example, Linux users likely will prefer the version that is installed via their package manager.

If you need to put some guard rails around it to make it so that y'all can still support the thing, I get it. Something like "This option is only for people who know what they're doing, and you should expect that we tell you to turn it off if you open a support request".

**Comment by achekulaev** [comment]
Docksal is not downloading any binary tools on Linux directly, get.docker.com script installs what is needed there.

The guardrails. "Do you really want to delete this file?" dialog window never stopped people from deleting the important files. So there will always be some guy, who heard from you, or from someone else about this option, and who used it when installed Docksal without understanding consequences, because someone told him that its safer, than having some wonky bash script that people just happen to use as a local development environment to download some random binaries for them. And then after some update that user is going to have weird issues, and tell people that Docksal is so unstable. 

I hope you understand why I have my reasons for push back on this, especially since I know for sure that those binaries are non-harmful, and that with high degree of probability you have dozens of random binaries lying around your filesystem right now. 

But I will keep this issue open and see if we get more people interested in the option, in which case I could add it.

**Comment by cweagans** [comment]
> I hope you understand why I have my reasons for push back on this

Yeah, I totally get it. Every extra option is just another thing to maintain, but this is really important and I don't think I'm effectively communicating the severity.

> especially since I know for sure that those binaries are non-harmful

I guess that's really the root of the issue. _You_ know that these binaries are safe, but the way that they were downloaded to my machine without confirmation or verification or anything doesn't give _me_ a lot of confidence that they are safe, especially since an _older_ version than what I already had was downloaded.

That's a really questionable outcome that should make pretty much anyone uncomfortable. Imagine if Docker had a security update and the tools that Docksal downloaded were older than that security update: suddenly, Docksal is responsible for downloading an unpatched version of software that has administrative access to their machine.

And for what? To avoid telling me to just update Docker if I'm having problems that you suspect might be Docker version related?

Just FYI: this issue alone is enough to turn me away from Docksal. I am simply not comfortable using it if downloading binaries like this is required. It's a weird solution to something that shouldn't be a problem in the first place.

**Comment by achekulaev** [comment]
> And for what? To avoid telling me to just update Docker if I'm having problems that you suspect might be Docker version related?

Because you are the only user of the app right...

**Comment by achekulaev** [comment]
There will be confirmation and hash verification in the next release, the ticket for this is #1260 linked above.

**Comment by cweagans** [comment]
lol ok. i'll find a different tool.

**Comment by attiks** [comment]
I mostly run the following

```
fin update --self
fin update --stack
fin update --system-images
```

**Comment by achekulaev** [comment]
@attiks so you also avoid tools download? Could you elaborate?

**Comment by attiks** [comment]
Running ubuntu, docker installed manually, so have to cancel it, but it seems that it still does some thing in the last step as well.

**Comment by achekulaev** [comment]
@attiks interesting. Generally you should not meet any issues since `install_tools_debian` function only runs Docker installation when `is_docker_version` or `is_docker_server_version` do not return satisfactory results. 
There might be a chance that you are running older `docker-compose` version, since this is a separate utility and is not installed by the `docker` package usually. If it is the case I owuld recommend updating to the version Docksal is tryin to install. 

Just for the sake of possibly debugging a detection issue, could you show output of 

```
docker version --format '{{.Client.Version}}'
docker version --format '{{.Server.Version}}'
docker-compose version --short 
```

**Comment by attiks** [comment]
I get this when running `fin project start`

 ALERT:  Required Docker server version is 19.03.4

```bash
peter@bto003:~$ docker version --format '{{.Client.Version}}'
19.03.3
peter@bto003:~$ docker version --format '{{.Server.Version}}'
19.03.3
peter@bto003:~$ docker-compose version --short 
1.21.0
```

`docker-ce is already the newest version (5:19.03.3~3-0~ubuntu-disco).`

**Comment by achekulaev** [comment]
Oh I see... Interesting issue... They just don't have 19.04 package. 
This is a nuisance, but why do you cancel Docker installation? Is that because you worry that installer script might break you current one?

**Comment by attiks** [comment]
> Is that because you worry that installer script might break you current one?

Yes, that's what the message implies ;-)

**Comment by achekulaev** [comment]
Oh yes right, that get.docker.com script indeed has this message. THis is a very valid use case. I will reopen this ticket to think more on this.

**Comment by fjmk** [comment]
@attiks I have Kubuntu 18.04 (LTS) version, latest ubuntu docker.io version is 18.09.7-0ubuntu1~18.04.4.
To solve my problem I used the installation instructions [here](https://docs.docker.com/install/linux/docker-ce/ubuntu/)  This makes apt-get upgrade updating always to the latest version. I didn't have any problems since I installed it 18 months ago.

PS: I don't know if 19.04-docker.io is still using the aufs driver. If yes, you have two choices:
1. Configure the new installation with aufs if you want to keep the old images & containers
2. Or  stop docker, delete /var/lib/docker/* and restart docker. Docksal will download the images again :)

**Comment by ms-marcus** [comment]
I also have this use-case (and run similar commands to @attiks). I run docksal as part of a CI system on a range of servers managed by puppet (https://github.com/microserve-io/puppet-docksal/blob/master/manifests/config.pp). I didn't want docksal downloading and overwriting the docker services that puppet maintains. 

Again, probably a fairly minor use-case and I've got a workaround so I'm not desperate but would still be a nice-to-have for my use-case :smile:
